File: typescript/libs/workspace/adapters/system/src/lib/session/originstate/DatabaseOriginState.ts
import { Operations } from '@allors/workspace/domain/system';
import { MethodType, RoleType } from '@allors/workspace/meta/system';
import { DatabaseRecord } from '../../database/DatabaseRecord';
import { IRecord } from '../../IRecord';
import { Strategy } from '../Strategy';
import { RecordBasedOriginState } from './RecordBasedOriginState';

export const UnknownVersion = 0;
export const InitialVersion = 1;

export abstract class DatabaseOriginState extends RecordBasedOriginState {
  DatabaseRecord: DatabaseRecord;

  protected constructor(strategy: Strategy, record: DatabaseRecord) {
    super(strategy);
    this.DatabaseRecord = record;
    this.PreviousRecord = this.DatabaseRecord;
  }

  public get Version(): number {
    return this.DatabaseRecord?.version ?? UnknownVersion;
  }

  private get IsVersionUnknown(): boolean {
    return this.Version == UnknownVersion;
  }

  protected get ExistDatabaseRecord(): boolean {
    return this.Record != null;
  }

  get RoleTypes(): Set<RoleType> {
    return this.Class.databaseOriginRoleTypes;
  }

  get Record(): IRecord {
    return this.DatabaseRecord;
  }

  public CanRead(roleType: RoleType): boolean {
    if (!this.ExistDatabaseRecord) {
      return true;
    }

    if (this.IsVersionUnknown) {
      return false;
    }

    const permission = this.Session.workspace.database.getPermission(this.Class, roleType, Operations.Read);
    return this.DatabaseRecord.isPermitted(permission);
  }

  public CanWrite(roleType: RoleType): boolean {
    if (!this.ExistDatabaseRecord) {
      return true;
    }

    if (this.IsVersionUnknown) {
      return false;
    }

    const permission = this.Session.workspace.database.getPermission(this.Class, roleType, Operations.Write);
    return this.DatabaseRecord.isPermitted(permission);
  }

  public CanExecute(methodType: MethodType): boolean {
    if (!this.ExistDatabaseRecord) {
      return true;
    }

    if (this.IsVersionUnknown) {
      return false;
    }

    const permission = this.Session.workspace.database.getPermission(this.Class, methodType, Operations.Execute);
    return this.DatabaseRecord.isPermitted(permission);
  }

  public PushResponse(newDatabaseRecord: DatabaseRecord) {
    this.DatabaseRecord = newDatabaseRecord;
    this.ChangedRoleByRelationType = null;
  }

  public OnPulled() {
    this.DatabaseRecord = this.Session.workspace.database.getRecord(this.Id);
  }

  OnChange() {
    this.Session.changeSetTracker.OnDatabaseChanged(this);
    this.Session.pushToDatabaseTracker.OnChanged(this);
  }
}

File: dotnet/System/workspace/csharp/adapters/allors.workspace.adapters/session/originstate/DatabaseOriginState.cs
// <copyright file="DatabaseOriginState.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters
{
    using System.Collections.Generic;
    using Meta;

    public abstract class DatabaseOriginState : RecordBasedOriginState
    {
        protected DatabaseOriginState(DatabaseRecord record)
        {
            this.DatabaseRecord = record;
            this.PreviousRecord = this.DatabaseRecord;
        }

        public long Version => this.DatabaseRecord?.Version ?? Allors.Version.Unknown;

        internal bool IsVersionUnknown => this.Version == Allors.Version.Unknown.Value;

        protected bool ExistDatabaseRecord => this.Record != null;

        protected override IEnumerable<IRoleType> RoleTypes => this.Class.DatabaseOriginRoleTypes;

        protected override IRecord Record => this.DatabaseRecord;

        protected DatabaseRecord DatabaseRecord { get; set; }

        public bool CanRead(IRoleType roleType)
        {
            if (!this.ExistDatabaseRecord)
            {
                return true;
            }

            if (this.IsVersionUnknown)
            {
                return false;
            }

            var permission = this.Session.Workspace.DatabaseConnection.GetPermission(this.Class, roleType, Operations.Read);
            return this.DatabaseRecord.IsPermitted(permission);
        }

        public bool CanWrite(IRoleType roleType)
        {
            if (!this.ExistDatabaseRecord)
            {
                return true;
            }

            if (this.IsVersionUnknown)
            {
                return false;
            }

            var permission =
                this.Session.Workspace.DatabaseConnection.GetPermission(this.Class, roleType, Operations.Write);
            return this.DatabaseRecord.IsPermitted(permission);
        }

        public bool CanExecute(IMethodType methodType)
        {
            if (!this.ExistDatabaseRecord)
            {
                return true;
            }

            if (this.IsVersionUnknown)
            {
                return false;
            }

            var permission =
                this.Session.Workspace.DatabaseConnection.GetPermission(this.Class, methodType, Operations.Execute);
            return this.DatabaseRecord.IsPermitted(permission);
        }

        public void PushResponse(DatabaseRecord newDatabaseRecord)
        {
            this.DatabaseRecord = newDatabaseRecord;
            this.ChangedRoleByRelationType = null;
        }
        public void OnPulled() =>
            // TODO: check for overwrites
            this.DatabaseRecord = this.Session.Workspace.DatabaseConnection.GetRecord(this.Id);

        protected override void OnChange()
        {
            this.Session.ChangeSetTracker.OnDatabaseChanged(this);
            this.Session.PushToDatabaseTracker.OnChanged(this);
        }
    }
}